log4j-internal-util/src/main/java/org/apache/logging/log4j/internal/util/changelog/exporter/AsciiDocExporter.java

```diff
@@ -0,0 +1,424 @@
+
+                        // Skip the directory itself.
+                        if (path.equals(directory)) {
+                            return false;
+                        }
+
+                        // Skip hidden files.
+                        boolean hiddenFile = path.getFileName().toString().startsWith(".");
+                        if (hiddenFile) {
+                            return false;
+                        }
+
+                        // Accept the rest.
+                        return true;
+
+                    });
+        } catch (final IOException error) {
+            final String message = String.format("failed walking directory: `%s`", directory);
+            throw new UncheckedIOException(message, error);
+        }
+    }
+
+    private static void exportRelease(
+            final Path projectRootDirectory,
+            final Path releaseDirectory,
+            final ChangelogRelease changelogRelease) {
+        final String introAsciiDoc = readIntroAsciiDoc(releaseDirectory);
+        final List<ChangelogEntry> changelogEntries = readChangelogEntries(releaseDirectory);
+        try {
+            exportRelease(projectRootDirectory, changelogRelease, introAsciiDoc, changelogEntries);
+        } catch (final IOException error) {
+            final String message = String.format("failed exporting release from directory `%s`", releaseDirectory);
+            throw new UncheckedIOException(message, error);
+        }
+    }
+
+    private static String readIntroAsciiDoc(final Path releaseDirectory) {
+
+        // Determine the file to be read.
+        final Path introAsciiDocFile = ChangelogFiles.introAsciiDocFile(releaseDirectory);
+        if (!Files.exists(introAsciiDocFile)) {
+            return "";
+        }
+
+        // Read the file.
+        final List<String> introAsciiDocLines;
+        try {
+            introAsciiDocLines = Files.readAllLines(introAsciiDocFile, StandardCharsets.UTF_8);
+        } catch (final IOException error) {
+            final String message = String.format("failed reading intro AsciiDoc file: `%s`", introAsciiDocFile);
+            throw new UncheckedIOException(message, error);
+        }
+
+        // Erase comment blocks.
+        final boolean[] inCommentBlock = {false};
+        return introAsciiDocLines
+                .stream()
+                .filter(line -> {
+                    final boolean commentBlock = "////".equals(line);
+                    if (commentBlock) {
+                        inCommentBlock[0] = !inCommentBlock[0];
+                        return false;
+                    }
+                    return !inCommentBlock[0];
+                })
+                .collect(Collectors.joining("\n"))
+                + "\n";
+
+    }
+
+    private static List<ChangelogEntry> readChangelogEntries(final Path releaseDirectory) {
+        return findAdjacentFiles(releaseDirectory)
+                .sorted(FILE_MODIFICATION_TIME_COMPARATOR)
```

Comment:
Are we sure that `git clone` sets the correct modification time?

Reply:
That is a really good point @ppkarwasz! I don't know and I think it indeed not a good idea to rely on that. I have fixed this in d64770b1ca056a69b472759a1048a340838d83fc by ordering changelog entries by filepath.